File: c++/leetcode/367. Valid Perfect Square.cpp
class Solution
{
public:
  bool isPerfectSquare(int num)
  {
    int start = 1;
    int end = num;

    while (start <= end)
    {
      long long mid = start + ((end - start) / 2);

      auto sqr = mid * mid;

      if (sqr == num)
        return true;

      if (sqr > num)
        end = --mid;
      else
        start = ++mid;
    }

    return false;
  }
};

File: javascript/leetcode/367. Valid Perfect Square.js
/**
 * @param {number} num
 * @return {boolean}
 */
var isPerfectSquare = function (num) {
  let start = 1;
  let end = num;

  while (start <= end) {
    let mid = start + parseInt((end - start) / 2);
    const sqr = mid * mid;

    if (sqr == num) return true;
    if (sqr > num) end = --mid;
    else start = ++mid;
  }

  return false;
};
